### PR TITLE
Add support for generating Azure Blob Storage signed URLs

### LINKED ISSUE #18: Microsoft Azure backed Delta tables
### Contact Details

ranko.bacevic@dev.telnet.rs

### Is your feature request related to a problem? Please describe.

I was trying to connect sharing server with ADLS gen 2, but I am getting internal server error once I execute request.

### Describe the solution you'd like.

Hi,
is supporting Microsoft Azure backed Delta tables in development? Is there some estimate when it is expected to be done, or if help is needed, I can try to help, with some minor guidelines and info from someone in charge :) ?
Thanks in advance!

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by tdikland**
This PR introduces support for sharing Delta Lake tables stored on Azure Blob Storage. 

I've marked the PR as WIP, since I feel there are still a couple of open questions that we may want to address in this or subsequent issues.

1. Is it okay to change the signature of the function signing the urls to be async?
The crate that I used to call the Azure SDK for building a pre-signed URL has an async method to do so. I figured this was okay, since all route handlers are async anyway.  
2. What should we do if there are no credentials available to sign the urls?
Right now the urls are passed through to the client unsigned. We may want to present an error instead.
3. In Azure there are multiple possibilities for authentication to Azure Blob Storage. I've chosen to implement only the Shared Key method, since that brings this crate at parity with the scala implementation

On a higher level:
1. I think it is a good idea to bring in a crate that can handle signing URLs for the different cloud object stores where the Delta Table lives. To this end I've published a crate that I used in some other personal projects to do exactly this `https://crates.io/crates/cloud-file-signer`. This crate defines a `CloudFileSigner` trait and implements it for a couple of popular object stores (S3, Azure Blob Storage, GCS) and could be extended to other stores (R2, MinIO, etc.)

Looking forward to some feedback!

Closes #18

### CODE REVIEW

**Comment by ognis1205** [comment]
Hi Tim, thank you for submitting this PR!

I haven't delved into the details of the PR yet, but it looks great. That aside, I like your idea of the Delta Sharing Server SDK, the delta-sharing-server-rs project. Since the Delta Sharing protocol is defined through REST APIs, there may be some considerations to keep in mind; nevertheless, the idea is interesting.

As for the cloud-file-signer crate, I believe it would be more beneficial to expand it into a larger project, similar to the delta-sharing-server-rs. If your contribution here serves any purpose, please feel free to incorporate it.

**Comment by ognis1205** [comment]
@tdikland 

Regarding the open questions, here are my answers:

1. Yes, it is totally fine, and it would be better to do so.
2. According to [the official protocol specification](https://github.com/delta-io/delta-sharing/blob/main/PROTOCOL.md#read-data-from-a-table), there is no description. However, I think we should return a 500 error and trace the error.
3. For now, I would not like to introduce complexity, so I really appreciate your implementation that aligns with the reference server implementation. When we want to add further options, we should create another issue.

**Comment by tdikland** [comment]
Thanks for your quick reaction and enthusiasm @ognis1205. 

Regarding the `delta-sharing-server-rs` project, I think it would be great for users to have both the lib that implements the delta sharing protocol and the batteries included version (which this crate is currently focusing on). The user journey that I'm thinking about is a user starts by deploying the reference implementation of the delta sharing server and when customisation is needed they can take the lib and plugin custom auth, retry, other middleware, different database etc. It may be nice to support this user journey with this project. I will open a new issue to discuss this in more depth.

Regarding the support for the underlying object stores (i.e. AWS, Azure, GCP). Thanks for your insights on my questions. I will be iterating on the implementation and let you know when it's ready for review.

**Comment by ognis1205** [comment]
I am very grateful for your contribution @tdikland !

This is a quick review of the current commit. Regarding the 'cloud-file-signer' crate, I believe there is operational experience in your project, but I think it still needs a bit more incubation. Therefore, I believe it's still early to use the crate. Since this project itself is also in incubation, considering the future development speed, I would prefer to have the implementation of signed URLs for each cloud provider done here as much as possible. The experience gained here can then be reflected in 'cloud-file-signer,' and eventually, I would like to use the crate in 'delta-sharing-rs' as well.

Please do not remove `utilities::signed_url`. Instead, can we consider implementing a policy to instantiate the cloud signer instance from the credentials of each provider using a factory method (similar to what is done in the `bootstrap` module)? I would like a factory method to be added to the `struct Utility`. The reason is that having information in the form of logging, indicating which provider's credentials were loaded during server bootstrap, would be more informative and would ensure compatibility with other modules. What do you think?

**Comment by ognis1205** [comment]
@tdikland 

Even if, for instance, we are not currently using the crate, I am considering stating in the README that we anticipate potential future use of it as a related project.

**Comment by tdikland** [comment]
Thanks for the feedback. I agree it is not yet desirable to depend on the `cloud-file-signer` crate. I was experimenting with it and accidentally pushed it to this PR (sorry!). I'll give a heads-up once it is ready for review!

**Comment by ognis1205** [comment]
You don't have to apologize for it at all! I apologize if I have caused any disruption to the development. We should have prepared a more accessible system, such as a Code of Conduct (COC), to make participation easier. It was very helpful that you made various suggestions and asked questions.

**Comment by tdikland** [comment]
Allright, I think I got it down now.

**Comment by ognis1205** [comment]
Thank you for your cooperation @tdikland :)

I believe we have reached a consensus on the general direction of the implementation. It seems there are still some parts that I would like to discuss though, before proceeding to the review, let's go ahead and run some tests for now. @roeap could you please approve the workflow?

**Comment by tdikland** [comment]
Hi @ognis1205, @roeap, I'm ready for reviewing.

File: src/server/utilities/signed_url.rs [hunk 1]
```diff
@@ -38,146 +32,200 @@ impl FromStr for Platform {
...
-            "gs" => Ok(Self::Gcp {
-                url: String::from(url.as_str()),
-                bucket: String::from(url.domain().unwrap_or("")),
-                path: String::from(url.path().strip_prefix('/').unwrap_or("")),
-            }),
-            _ => Ok(Self::None {
-                url: String::from(url.as_str()),
-            }),
+            "s3" | "s3a" => Ok(Self::Aws),
+            "abfss" | "abfs" => Ok(Self::Azure),
```
Comment (roeap) [inline src/server/utilities/signed_url.rs:36]: had interesting discussions around this in the past :D - essentially in azure there are many ways a scheme could look like, some "official" some de-facto standards (e.g. what `fsspec` adopted).. Also onelake did not make things less confusing.

Technically `abfs[s]` refers to a specific implementation of an hdfs compliant driver...

In the end its fine to do this for now, I'll eave some comments later around the general signing question.

**Review by roeap — COMMENTED** [review]
> I think it is a good idea to bring in a crate that can handle signing URLs for the different cloud object stores where the Delta Table lives. To this end I've published a crate that I used in some other personal projects to do exactly this https://crates.io/crates/cloud-file-signer. This crate defines a CloudFileSigner trait and implements it for a couple of popular object stores (S3, Azure Blob Storage, GCS) and could be extended to other stores (R2, MinIO, etc.)

In the larger rust-arrow ecosystem the community has been developing the `object_store` crate which abstracts access to a large number of object stores, without depending on all the individual SDKs to avoid bloat.

@tustvold - would adding methods to acquire pre-signed URls something you see in-scope for object store, or would that dilute the focussed API too much?

I did not see the discussion around `delta-sharing-server-rs`, but in the past @ognis1205 and I had discussed that we want to break this up into smaller crates. and split the different "sub-apis" such that one can easily mix-and-match a server that supports the sharing protocol with just the desired features.

Overall looking great - If @ognis1205 gives his OK, I am happy to merge - knowing that we have some way to go before we can release this project more "officially".

**Comment by tustvold** [comment]
> acquire pre-signed URls

We already support this for S3, Azure and GCP are more complex but would be happy for contributions in this space - https://docs.rs/object_store/latest/object_store/signer/trait.Signer.html

**Comment by ognis1205** [comment]
@tdikland @roeap 
I will review this PR in the next 12 hours, here in Tokyo, it's around 24:00 midnight. 

> I did not see the discussion around delta-sharing-server-rs, but in the past @ognis1205 and I had discussed that
> we want to break this up into smaller crates. and split the different "sub-apis" such that one can easily mix-and-match
> a server that supports the sharing protocol with just the desired features.

Is there any place to discuss this issue? I have some opinions on this since the protocol is defined over REST APIs.

**Comment by tustvold** [comment]
FWIW I've created two upstream tickets on object_store for adding support for Azure and GCP, along with links to the relevant docs. I think they should be relatively straightforward if anyone wanted to pick them up:

* https://github.com/apache/arrow-rs/issues/5232
* https://github.com/apache/arrow-rs/issues/5233

**Comment by roeap** [comment]
@tustvold  - i know the azure apis quite well, and signing using keys should (hopefully) be straignforward to leverage the existing signing using the master key ... user delegated are just API cal li think :) 

had a quick glance at GCP - might do that second - seems similar, cannonicalize request and sign ...

**Comment by roeap** [comment]
@tdikland - just FYI, started to add url signing for azure to object store https://github.com/apache/arrow-rs/pull/5259.

**Comment by tdikland** [comment]
@roeap @ognis1205 - FYI I have created two new issues to keep track of the discussion and next steps.
- https://github.com/delta-incubator/delta-sharing-rs/issues/23
- https://github.com/delta-incubator/delta-sharing-rs/issues/24

**Pull request merged**